File: engines/java/src/test/java/org/agentce/CoverageTest.java
package org.agentce;

import static org.junit.jupiter.api.Assertions.assertEquals;

import com.fasterxml.jackson.databind.JsonNode;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

/**
 * Coverage reconciliation is byte-identical to the reference (SPEC §6.5, §8.3): the {@code ok}/{@code
 * gap}/{@code unknown} roll-ups, repeating-decimal ratio rounding, and a manifest-declared denominator.
 */
class CoverageTest {

    @Test
    void coverageReconciliationMatchesGolden() {
        JsonNode fixture = Json.parseFile(TestPaths.testData().resolve("coverage-fixture.json"));
        JsonNode golden = Json.parseFile(TestPaths.testData().resolve("coverage-golden.json"));
        Profile profile = Profile.fromDict(fixture.get("profile"));
        JsonNode result = Coverage.computeCoverage(
                Fixtures.toList(fixture.get("events")), profile, TestPaths.testData().resolve("coverage-bundle"));
        assertEquals(Canonical.canonicalString(golden), Canonical.canonicalString(result));
    }

    private static Profile.CoverageDenominator denominator(String manifest) {
        Profile.CoverageDenominator denominator = new Profile.CoverageDenominator();
        denominator.kind = "registry";
        denominator.source = "urn:src:egress";
        denominator.covers = List.of("ToolCall");
        denominator.manifest = manifest;
        return denominator;
    }

    /** The denominator source's own ingested events: the fallback when no manifest is readable. */
    private static final Map<String, Map<String, Integer>> BY_SOURCE =
            Map.of("urn:src:egress", Map.of("ToolCall", 3));

    @Test
    void denominatorManifestEscapingTheBundleRootIsNotRead(@TempDir Path tempDir) throws IOException {
        Path bundleRoot = tempDir.resolve("bundle");
        Files.createDirectories(bundleRoot.resolve("reference"));
        Path outside = tempDir.resolve("secret.json");
        Files.writeString(outside, "{\"expected\": {\"ToolCall\": 999999}}", StandardCharsets.UTF_8);
        Files.createSymbolicLink(bundleRoot.resolve("reference").resolve("counts.json"), outside);

        // Both shapes of escape -- a literal '..' and an innocent-looking path that is a symlink out --
        // fall back to the source's own counts, exactly as a missing manifest does.
        assertEquals(
                Map.of("ToolCall", 3),
                Coverage.denominatorCounts(denominator("../secret.json"), BY_SOURCE, bundleRoot));
        assertEquals(
                Map.of("ToolCall", 3),
                Coverage.denominatorCounts(denominator("reference/counts.json"), BY_SOURCE, bundleRoot));
    }

    @Test
    void denominatorManifestInsideTheBundleRootIsRead(@TempDir Path tempDir) throws IOException {
        Path bundleRoot = tempDir.resolve("bundle");
        Files.createDirectories(bundleRoot.resolve("reference"));
        Files.writeString(
                bundleRoot.resolve("reference").resolve("counts.json"),
                "{\"expected\": {\"ToolCall\": 7}}",
                StandardCharsets.UTF_8);

        assertEquals(
                Map.of("ToolCall", 7),
                Coverage.denominatorCounts(denominator("reference/counts.json"), BY_SOURCE, bundleRoot));
    }
}

File: engines/typescript/src/coverage.test.ts
/**
 * Coverage reconciliation is byte-identical to the Python reference (SPEC §6.5, §8.3).
 *
 * `testdata/coverage-{fixture,golden}.json` were emitted by the reference engine's `compute_coverage`
 * over subjects that are covered, below threshold, and uncovered, exercising the `ok`/`gap`/`unknown`
 * roll-ups, repeating-decimal ratio rounding, and a manifest-declared denominator (read from
 * `testdata/coverage-bundle/`).
 */

import assert from "node:assert/strict";
import { mkdirSync, mkdtempSync, readFileSync, rmSync, symlinkSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { test } from "node:test";
import { canonicalString } from "./canonical";
import { computeCoverage } from "./coverage";
import { profileFromDict } from "./profile";

const TESTDATA = join(__dirname, "..", "testdata");

test("coverage reconciliation matches the Python reference golden", () => {
  const fixture = JSON.parse(readFileSync(join(TESTDATA, "coverage-fixture.json"), "utf-8"));
  const golden = JSON.parse(readFileSync(join(TESTDATA, "coverage-golden.json"), "utf-8"));
  const profile = profileFromDict(fixture.profile);
  const result = computeCoverage(fixture.events, profile, join(TESTDATA, "coverage-bundle"));
  assert.equal(canonicalString(result), canonicalString(golden));
});

// --- A denominator manifest is a bundle-adjacent reference: it must stay inside the bundle root
// --- after symlinks resolve, or it is not read at all (SPEC §6.5, §8.1).

const DENOMINATOR_SOURCE = "urn:src:egress";

/** One observed ToolCall plus two of the denominator source's own: the fallback expects 2. */
const EVENTS: Array<Record<string, unknown>> = [
  { source: "urn:src:agent", subject: "subj-A", data: { "@type": "ToolCall" } },
  { source: DENOMINATOR_SOURCE, subject: "subj-A", data: { "@type": "ToolCall" } },
  { source: DENOMINATOR_SOURCE, subject: "subj-A", data: { "@type": "ToolCall" } },
];

function profileWithManifest(manifest: string) {
  return profileFromDict({
    subjects: [
      {
        id: "subj-A",
        coverage_denominators: [
          { kind: "registry", source: DENOMINATOR_SOURCE, covers: ["ToolCall"], manifest },
        ],
      },
    ],
  });
}

/** The `expected` count the reconciliation used for `subj-A`'s ToolCall events. */
function expectedToolCalls(result: Record<string, unknown>): unknown {
  const subjects = result.subjects as Record<string, { event_types: Record<string, unknown> }>;
  const entry = subjects["subj-A"]?.event_types.ToolCall as { expected: unknown };
  return entry.expected;
}

function withTempBundle(body: (dir: string, root: string) => void): void {
  const dir = mkdtempSync(join(tmpdir(), "agentce-coverage-"));
  try {
    const root = join(dir, "bundle");
    mkdirSync(join(root, "reference"), { recursive: true });
    body(dir, root);
  } finally {
    rmSync(dir, { recursive: true, force: true });
  }
}

test("a denominator manifest escaping the bundle root with '..' is not read", () => {
  withTempBundle((dir, root) => {
    writeFileSync(join(dir, "secret.json"), '{"expected": {"ToolCall": 999999}}', "utf-8");

    const result = computeCoverage(EVENTS, profileWithManifest("../secret.json"), root);
    assert.equal(expectedToolCalls(result), 2); // the source's own ingested events, not the escape
  });
});

test("a denominator manifest whose symlink escapes the bundle root is not read", () => {
  withTempBundle((dir, root) => {
    const outside = join(dir, "secret.json");
    writeFileSync(outside, '{"expected": {"ToolCall": 999999}}', "utf-8");
    symlinkSync(outside, join(root, "reference", "counts.json"));

    const result = computeCoverage(EVENTS, profileWithManifest("reference/counts.json"), root);
    assert.equal(expectedToolCalls(result), 2);
  });
});

test("a denominator manifest inside the bundle root is still read", () => {
  withTempBundle((_dir, root) => {
    writeFileSync(join(root, "reference", "counts.json"), '{"expected": {"ToolCall": 7}}', "utf-8");

    const result = computeCoverage(EVENTS, profileWithManifest("reference/counts.json"), root);
    assert.equal(expectedToolCalls(result), 7);
  });
});

test("a denominator manifest that is a symlink inside the bundle root is still read", () => {
  withTempBundle((_dir, root) => {
    const real = join(root, "reference", "real-counts.json");
    writeFileSync(real, '{"counts": {"ToolCall": 5}}', "utf-8");
    symlinkSync(real, join(root, "reference", "counts.json"));

    const result = computeCoverage(EVENTS, profileWithManifest("reference/counts.json"), root);
    assert.equal(expectedToolCalls(result), 5);
  });
});
